app/models/spree/product_decorator.rb

```diff
@@ -18,6 +18,7 @@ def event_end_date
     end
 
     def event_discount
+      effective_flash_sale&.sale_products&.find_by(product_id: self.id)&.discount ||
```

Comment:
this will loop the query causing performance impact. I think it would be better to load eager load effective_sale_products from products ( custom association ). we no longer need to interact with active_sale_event directly.

Reply:
- add more fields necessary to the sale products
- add custom effective_sale_products
- make eager effective_sale_products to avoid n+1 query
- sync sale products with active sale events when changing
- validation when adding sale products, it should not allow a product added to effective sale events.